modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -0,0 +1,1655 @@
+        else if (constant instanceof Integer)   methodWriter.push((int)constant);
+        else if (constant instanceof Character) methodWriter.push((char)constant);
+        else if (constant instanceof Short)     methodWriter.push((short)constant);
+        else if (constant instanceof Byte)      methodWriter.push((byte)constant);
+        else if (constant instanceof Boolean)   methodWriter.push((boolean)constant);
+        else {
+            throw new IllegalStateException("unexpected constant [" + constant + "]");
+        }
+    }
+
+    @Override
+    public void visitNull(NullNode irNullNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.visitInsn(Opcodes.ACONST_NULL);
+    }
+
+    @Override
+    public void visitDefInterfaceReference(DefInterfaceReferenceNode irDefInterfaceReferenceNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irDefInterfaceReferenceNode.getLocation());
+
+        // place holder for functional interface receiver
+        // which is resolved and replace at runtime
+        methodWriter.push((String)null);
+
+        for (String capture : irDefInterfaceReferenceNode.getCaptures()) {
+            WriteScope.Variable variable = writeScope.getVariable(capture);
+            methodWriter.visitVarInsn(variable.getAsmType().getOpcode(Opcodes.ILOAD), variable.getSlot());
+        }
+    }
+
+    @Override
+    public void visitTypedInterfaceReference(TypedInterfaceReferenceNode irTypedInterfaceReferenceNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irTypedInterfaceReferenceNode.getLocation());
+
+        for (String capture : irTypedInterfaceReferenceNode.getCaptures()) {
+            WriteScope.Variable variable = writeScope.getVariable(capture);
+            methodWriter.visitVarInsn(variable.getAsmType().getOpcode(Opcodes.ILOAD), variable.getSlot());
+        }
+
+        methodWriter.invokeLambdaCall(irTypedInterfaceReferenceNode.getReference());
+    }
+
+    @Override
+    public void visitTypeCaptureReference(TypedCaptureReferenceNode irTypedCaptureReferenceNode, WriteScope writeScope) {
+        MethodWriter methodWriter = writeScope.getMethodWriter();
+        methodWriter.writeDebugInfo(irTypedCaptureReferenceNode.getLocation());
+        Variable captured = writeScope.getVariable(irTypedCaptureReferenceNode.getCaptures().get(0));
+
+        methodWriter.visitVarInsn(captured.getAsmType().getOpcode(Opcodes.ILOAD), captured.getSlot());
+        Type methodType = Type.getMethodType(MethodWriter.getType(irTypedCaptureReferenceNode.getExpressionType()), captured.getAsmType());
+        methodWriter.invokeDefCall(irTypedCaptureReferenceNode.getMethodName(),
+                methodType, DefBootstrap.REFERENCE, irTypedCaptureReferenceNode.getExpressionCanonicalTypeName());
```

Comment:
These got pretty long, pull some variables out.

Reply:
Split the lines to make it clearer.